keras_cv/models/backbones/densenet/densenet_backbone.py

```diff
@@ -0,0 +1,369 @@
+    )(x)
+    x = layers.Activation("relu", name=f"{name}_1_relu")(x)
+    x = layers.Conv2D(
+        growth_rate,
+        3,
+        padding="same",
+        use_bias=False,
+        name=f"{name}_2_conv",
+    )(x)
+    x = layers.Concatenate(axis=BN_AXIS, name=f"{name}_concat")([shortcut, x])
+    return x
+
+
+@keras.utils.register_keras_serializable(package="keras_cv.models")
+class DenseNetBackbone(Backbone):
+    """Instantiates the DenseNet architecture.
+
+    Args:
+        stackwise_num_repeats: list of ints, number of repeated convolutional
+            blocks per dense block.
+        include_rescaling: bool, whether to rescale the inputs. If set
+            to `True`, inputs will be passed through a `Rescaling(1/255.0)`
+            layer.
+        input_shape: optional shape tuple, defaults to (None, None, 3).
+        input_tensor: optional Keras tensor (i.e. output of `layers.Input()`)
+            to use as image input for the model.
+        compression_ratio: float, compression rate at transition layers.
+        growth_rate: int, number of filters added by each dense block.
+
+    Examples:
+    ```python
+    input_data = tf.ones(shape=(8, 224, 224, 3))
+
+    # Pretrained backbone
+    model = keras_cv.models.DenseNetBackbone.from_preset("densenet121_imagenet")
+    output = model(input_data)
+
+    # Randomly initialized backbone with a custom config
+    model = DenseNetBackbone(
+        stackwise_num_repeats=[6, 12, 24, 16],
+        include_rescaling=False,
+    )
+    output = model(input_data)
+    ```
+    """  # noqa: E501
+
+    def __init__(
+        self,
+        *,
+        stackwise_num_repeats,
+        include_rescaling,
+        input_shape=(None, None, 3),
+        input_tensor=None,
+        compression_ratio=0.5,
+        growth_rate=32,
+        **kwargs,
+    ):
+        inputs = utils.parse_model_inputs(input_shape, input_tensor)
+
+        x = inputs
+        if include_rescaling:
+            x = layers.Rescaling(1 / 255.0)(x)
+
+        x = layers.Conv2D(
+            64, 7, strides=2, use_bias=False, padding="same", name="conv1/conv"
+        )(x)
+        x = layers.BatchNormalization(
+            axis=BN_AXIS, epsilon=BN_EPSILON, name="conv1/bn"
+        )(x)
+        x = layers.Activation("relu", name="conv1/relu")(x)
+        x = layers.MaxPooling2D(3, strides=2, padding="same", name="pool1")(x)
+
+        pyramid_level_inputs = {}
+        for stack_index in range(len(stackwise_num_repeats) - 1):
+            index = stack_index + 2
+            x = apply_dense_block(
+                x,
+                stackwise_num_repeats[stack_index],
+                growth_rate,
+                name=f"conv{index}",
+            )
+            x = apply_transition_block(
+                x, compression_ratio, name=f"pool{index}"
+            )
+            pyramid_level_inputs[index] = x.node.layer.name
+
+        x = apply_dense_block(
```

Comment:
I don't think this works. First, we have hard coded `3` as the last index. Second, it seems like you need to iterate to `len(stackwise_num_repeats) - 2` to avoid hitting this twice, no?

I think the cleanest way might be to check for the last index in the loop. Something like
```python
if stack_index < len(stackwise_num_repeats) - 1:
    x = apply_transition_block(
        x, compression_ratio, name=f"pool{index}"
    )
```

Reply:
I doubt `DenseNetPresetSmokeTest` passes. You want to run `large` tests locally to check.

Reply:
The range function doesn't take into consideration the number it itself holds, so if it's `range(4 - 1)`, it takes `stack_index` as 0, 1, 2. Regrettably, the local testing error for me still persists. We might want to /gcbrun this once here.

For the loop, I'm going for another approach where the dense block is applied individually after it ends rather than the if condition being checked in every iteration. If the latter seems cleaner, do let me know.

Reply:
Seems cleaner, thank you!